File: dataset/data2/NMF.m
function [P] = NMF(A,lncSimi,disSimi,beita,k,iterate)
    %ALP=A;
    A=load('LD_adjmat.txt');
    load('RNA_s_kernel-k=10-t=10.mat', 'WM');
    lncSimi = WM;
    load('diease_similarity_kernel-k=10-t=10.mat', 'WM');
    disSimi = WM;
    beita = 0.01;
    k = 64;
    iterate = 500;
    [rows,cols]=size(A);%[585*88]
    % 初始化circRNA和disease迭代矩阵
%     这行代码在MATLAB中的意思是生成一个大小为rows x k的随机矩阵C，
%     矩阵中的元素是从[0,1)范围内均匀分布的随机数，
%     然后对矩阵中的每个元素取绝对值（abs函数）。
%     换句话说，该代码生成了一个非负的随机矩阵，其中每个元素都是0到1之间的随机数。
    C = abs(rand(rows,k));
    D = abs(rand(cols,k));
    %[test_results(fold), ~] = improve_NMFLP(X_test,circSimi,disSimi,0.01,0.7,300);
    % 计算对角矩阵
    diag_cf=diag(sum(lncSimi,2));
    diag_df=diag(sum(disSimi,2));
     % 避免假阴性数据对矩阵分解的影响，因此对邻接矩阵进行重构
    A1=lncSimi*A;
    A2=A*disSimi;
    for i=1:rows
        for j=1:cols
            A(i,j)=max(A1(i,j),A2(i,j));
        end
    end
    
    % 归一化
    AA=zeros(rows,cols);
    for j =1:cols
        colList=A(:,j);
        for i =1:rows
        AA(i,j)=(A(i,j)-min(colList))/(max(colList)-min(colList));
        end
    end
    A=AA;
    
    % 鲁棒性矩阵分解部分
    for step = 1:iterate
        Y=A-C*D';
        B=L21_norm(Y);
        
        % for circRNA coding
        if  beita >0
            BAD=B*A*D+beita*lncSimi*C;
            BCDD=B*C*(D')*D+beita*diag_cf*C;
        end
        C=C.*(BAD./BCDD);
        
        %for disease coding
        if  beita >0
            ABC=(A')*B*C+beita*disSimi*D;
            DCBC=D*(C')*B*C+beita*diag_df*D;
        end
        D=D.*(ABC./DCBC);
        
        scoreMat_NMF=A-C*D';
        error=mean(mean(abs(scoreMat_NMF)))/mean(mean(A));
        fprintf('step=%d  error=%f\n',step,error);
    end
     save C C;
     save D D;

File: src/NMF.py
# -*- coding: utf-8 -*-
"""
Created on Thu Nov 30 12:06:56 2023

@author: 28473
"""
from scipy.io import loadmat
import numpy as np
import torch
import numpy as np

def l21_norm(Mat):
    rows, cols = Mat.shape
    diagList = np.zeros(rows)

    for i in range(rows):
        add = np.sum(Mat[i, :]**2)
        diagList[i] = 1 / (2 * np.sqrt(add))

    diagTemp = np.diag(diagList)
    return diagTemp


def nmf(A, lncSimi, disSimi, beita, k, iterate):
    rows, cols = A.shape
    C = np.abs(np.random.rand(rows, k))
    D = np.abs(np.random.rand(cols, k))

    diag_cf = np.diag(np.sum(lncSimi, axis=1))
    diag_df = np.diag(np.sum(disSimi, axis=1))

    A1 = np.dot(lncSimi, A)
    A2 = np.dot(A, disSimi)
    A = np.maximum(A1, A2)

    AA = np.zeros_like(A)
    for j in range(cols):
        colList = A[:, j]
        AA[:, j] = (A[:, j] - np.min(colList)) / (np.max(colList) - np.min(colList))
    A = AA

    for step in range(iterate):
        Y = A - np.dot(C, D.T)
        B = l21_norm(Y)

        if beita > 0:
            BAD = np.dot(B, A) @ D + beita * lncSimi @ C
            BCDD = np.dot(B, C @ D.T @ D) + beita * diag_cf @ C
            C = np.multiply(C, BAD / BCDD)

        if beita > 0:
            ABC = A.T @ B @ C + beita * disSimi @ D
            DCBC = np.dot(D, C.T @ B @ C) + beita * diag_df @ D
            D = np.multiply(D, ABC / DCBC)

        scoreMat_NMF = A - np.dot(C, D.T)
        error = np.mean(np.abs(scoreMat_NMF)) / np.mean(A)
        #print(f'step={step}  error={error}')
    return C, D

# Example usage:
# if__name__="main"
# data_path = '../dataset/'
# data_set = 'data-R-D447-218/'
# A = np.loadtxt(data_path + data_set + 'disease-lncRNA.csv', delimiter=',')
# A=torch.tensor(A.T)
# disSimi1 = loadmat(data_path + data_set + 'diease_similarity_kernel-k=20-t=20-ALPHA=0.1.mat') 
# disSimi = disSimi1['WM']
# RNASimi1 = loadmat(data_path + data_set + 'RNA_s_kernel-k=20-t=20-ALPHA=0.1.mat')
# lncSimi = RNASimi1['WM']
# R,D=nmf(A, lncSimi, disSimi, 0.01, 64, 500)
# print(R.shape)
#print(D)
